app/models/subnet.rb

```diff
@@ -238,43 +269,51 @@ def self.parse_dhcp_options(options)
   end
 
   def validate_ranges
-    errors.add(:from, _("invalid IP address"))            if from.present? and !from =~ Net::Validations::IP_REGEXP
-    errors.add(:to, _("invalid IP address"))              if to.present?   and !to   =~ Net::Validations::IP_REGEXP
-    errors.add(:from, _("does not belong to subnet"))     if from.present? and not self.contains?(f=IPAddr.new(from))
-    errors.add(:to, _("does not belong to subnet"))       if to.present?   and not self.contains?(t=IPAddr.new(to))
-    errors.add(:from, _("can't be bigger than to range")) if from.present? and t.present? and f > t
     if from.present? or to.present?
       errors.add(:from, _("must be specified if to is defined"))   if from.blank?
       errors.add(:to,   _("must be specified if from is defined")) if to.blank?
     end
+    return if errors.keys.include?(:from) || errors.keys.include?(:to)
+    errors.add(:from, _("does not belong to subnet"))     if from.present? and not self.contains?(f=IPAddr.new(from))
+    errors.add(:to, _("does not belong to subnet"))       if to.present?   and not self.contains?(t=IPAddr.new(to))
+    errors.add(:from, _("can't be bigger than to range")) if from.present? and t.present? and f > t
+  end
+
+  def check_if_type_changed
+    if self.type_changed?
+      errors.add(:type, _("can't be updated after subnet is saved"))
+    end
   end
 
   def cleanup_addresses
-    self.network = cleanup_ip(network) if network.present?
-    self.mask = cleanup_ip(mask) if mask.present?
-    self.gateway = cleanup_ip(gateway) if gateway.present?
-    self.dns_primary = cleanup_ip(dns_primary) if dns_primary.present?
-    self.dns_secondary = cleanup_ip(dns_secondary) if dns_secondary.present?
+    IP_FIELDS.each do |f|
+      send("#{f}=", cleanup_ip(send(f))) if send(f).present?
+    end
     self
   end
 
   def cleanup_ip(address)
     address.gsub!(/\.\.+/, ".")
```

Comment:
Is this relevant for IPv6? I don't think they can be represented with dots so maybe this method should go to the subclass